### PR TITLE
feat: Introduce `heredoc_closing_marker` fixer

### PR DESCRIPTION
**Pull request by mvorisek**
unify CS by enforsing consistent heredoc/nowdoc closing marker

### CODE REVIEW

**Comment by mvorisek** [comment]
Very good point - https://www.jetbrains.com/help/phpstorm/using-language-injections.html.

I see value in unifying meaningless and usually very mixed `EOT`, `EOF`, ... markers into one like `EOD`.

I will update this PR to allow to specify a list of reserved markers which will be kept as is.

**Comment by Wirone** [comment]
@julienfalque That was exactly my impression when I took a quick look at the PR when I got notification. I agree 💯.

**Comment by mvorisek** [comment]
Reserved markers are supported and fixed to a normalized case.

**Comment by Wirone** [comment]
Before I take a deeper look at the implementation, what `EOD` stands for? End of ... 🤔? Doc? If yes, then I would rather see something more generic, because regardless of heredoc/nowdoc naming they're used for more scenarios, not only docs. Maybe `END` would be better, like in [official PHP docs](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc)?

**Comment by Wirone** [comment]
Also, I would consider standardising (by default, or behind an opt-in) heredoc to `<<<"END"` format (explicit `"`). Now string interpolation is implicit for `<<<END` and disabled for `<<<'END'` (nowdoc), and since explicit > implicit IMHO, then maybe it would be good to support enforcing it.

**Comment by mvorisek** [comment]
`EOD` stands for "end of data", it is quite common, `EOF` (end of file) and `EOT` (end of text) is used a lot of too.

**Comment by Wirone** [comment]
> `EOD` stands for "end of data", it is quite common, `EOF` (end of file) and `EOT` (end of text) is used a lot of too.

OK, thanks for clarifying 🙂. I would rather see `END` as default closing marker, but since this is configurable it does not matter that much. Up to you 😉.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [inline src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:104]:
```suggestion
                'Whether the closing marker should be wrapped in double quotes.'
```

**Review by Wirone — CHANGES_REQUESTED** [review]
@mvorisek This PR should be merged _after_ #7665 and we should ensure proper priorities, so `heredoc_closing_marker` is run after `multiline_string_to_heredoc`, because closing marker is not configurable there, and even if it would be, then there's no way to ensure both fixers are configured with the same marker (final output should respect closing marker and quotings from `heredoc_closing_marker`). I'm going to mark PR with "changes requested" to prevent accidental merge, but it does not mean it must be blocked - it's always possible to handle priorities later, after both fixers are merged, but I prefer to handle it here.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php]: done

**Comment by mvorisek** [comment]
> Also, I would consider standardising (by default, or behind an `explicit_string_interpolation` opt-in) heredoc to `<<<"END"` format (explicit `"`). Now string interpolation is implicit for `<<<END` and disabled for `<<<'END'` (nowdoc), and since explicit > implicit IMHO, then maybe it would be good to support enforcing it.

I like it, implemented.

About the `EOD` vs `END`, I belive `EOD` is better because it will less likely collide with real content - `end` is used in many programming languages.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (julienfalque) [inline src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:107]: I didn't even know the `<<<"FOO"` syntax was possible.

`<<<FOO` seems to be the most widely used syntax. PER Coding Styles does not give requirements about this but examples only use the implicit syntax. Shouldn't we set the default value to `false`?

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (julienfalque) [inline src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:100]: This list will never be always accurate, while you only give a few examples of markers that should be replaced (`EOT` and `EOF`). Instead of excluding markers, I think it would be easier to give the ones to replace. We could even merge options `reserved_closing_markers` and `closing_marker` so that users can provide their own replacement map, e.g.:
```php
[
    'EOT' => 'EOD',
    'EOF' => 'EOD',
]
```

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:107]: Personally I would go with implicit too (no `"`), even though I prefer explicit format. We can always add this fixer to `@PhpCsFixer` with proper config, but for wider audience I believe `<<<EOD` is more fitting, as official docs mentions `"` quotings for heredoc somewhere at the end, as optional.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:100]: I am not sure... I believe there _is_ some kind of closed set of functional endings, outside of which any other ending should be normalised. The difference in the approach is that with @mvorisek's you will get the diff applied/displayed and you must make decision if you should exclude ending from fixer or accept the change, while in @julienfalque non-listed endings won't be covered until someone discovers they exist in the codebase and adds mapping for them. Personally I prefer the former, because it enforces explicit decision over any found ending marker.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php]: 👍, implemented.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:100]: Let's keep the "reserved list". In a typical project, there are typically not many, very often zero, additional reserved closing markers needed outside the one predefined.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [inline src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:55]:
```suggestion
                        <?php $a = <<<TEST
```

Code sample should show that `"` is added when `explicit_heredoc_style` is enabled.

Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php]: 👍, done.

**Comment by mvorisek** [comment]
> @mvorisek This PR should be merged _after_ #7665 and we should ensure proper priorities, so `heredoc_closing_marker` is run after `multiline_string_to_heredoc`, because closing marker is not configurable there, and even if it would be, then there's no way to ensure both fixers are configured with the same marker (final output should respect closing marker and quotings from `heredoc_closing_marker`). I'm going to mark PR with "changes requested" to prevent accidental merge, but it does not mean it must be blocked - it's always possible to handle priorities later, after both fixers are merged, but I prefer to handle it here.

Let's land this PR and solve the priorities /w #7665 in that PR because of https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7665#issuecomment-1875484219. This PR itself has no priority conflict with the latest master.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [inline src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:29]: Don't you want to include `@author` information? PS. I still need to review actual implementation, but I'll do this later.

Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:32]: Thank you - I added my name and link to my GH profile.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:117]: Closing markers are case sensitive - https://3v4l.org/njpC9 - so in theory, some developer may want/need closing marker not in a upper case.

> use multibyte version of strtoupper

done

File: tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php
Comment (mvorisek) [reply tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php:262]: see https://3v4l.org/98K7e - automatic conflict resolution, very powerful, as then all heredocs can be fixed, even the one with otherwise conflicting markers, nevertheless such conflicts are very rare in real strings

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:103]: If there is no API to extract option default value from a rule, yes, it should be added, but not related with this PR :)

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php]: PHP JIT might optimize the mb_strtolower calls in the loop, but opcache itself cannot :)

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:103]: I am closing this thread as resolved, feel free to create feature request from it.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:117]: Yes, but uppercase markers would be only used for comparison, then you would fetch the marker from the config with desired casing (by determined index).

File: tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php
Comment (mvorisek) [reply tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php:1]: `EXPECTED` + `INPUT` no, who will maintain it? When you develop testcases, you very often copy between these. See https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/commit/481358c52db235f6e4840e55da1853b6de53ea38, this is the proper answer and fully supported by this fixer ;-)

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (Wirone) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php:103]: There's no way to get defaults from the config or extend them instead of overriding. I would like to see this in introducing PR, because it's a minute of work on existing branch (new PR requires more work and other maintainer's approval).

File: tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php
Comment (Wirone) [reply tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php:1]: `PHP` is good enough 👍. But I'm pretty sure there are mentioned markers in our codebase 😉.

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php]: quite atypical, searched `public const.*\[`, but done

File: src/Fixer/StringNotation/HeredocClosingMarkerFixer.php
Comment (mvorisek) [reply src/Fixer/StringNotation/HeredocClosingMarkerFixer.php]: done

File: tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php
Comment (Wirone) [reply tests/Fixer/StringNotation/HeredocClosingMarkerFixerTest.php:262]: Thanks for the explanation 🙂.

**Comment by Wirone** [comment]
Thank you very much @mvorisek 🍻!

**Pull request merged**
